BedrockServer.cpp

```diff
@@ -1351,107 +1369,91 @@ void BedrockServer::prePoll(fd_map& fdm) {
+            SINFO("[performance] network thread timing: "
+                << chrono::duration_cast<chrono::milliseconds>(end - lastStartTime).count() << " ms total elapsed. "
+                << chrono::duration_cast<chrono::milliseconds>(busyTime).count() << " ms busy. "
+                << chrono::duration_cast<chrono::milliseconds>(waitTime).count() << " ms waiting.");
+
+            // Reset everything.
+            waitTime = chrono::steady_clock::duration::zero();
+            lastStartTime = end;
         }
-        // See if the two iterators have the same FD.
-        if (fdmIt->first == (*socketSetIt)->s) {
-            // They do! We want to keep this, and then move on with both of them.
-            sockets.insert(*socketSetIt);
-            fdmIt++;
-            socketSetIt++;
-        } else if (fdmIt->first < (*socketSetIt)->s) {
-            // So, if the iterator into FDM is less than the iterator into socket set, we can discard it, all the
-            // values in socketSet are larger than this.
-            fdmIt++;
-        } else if ((*socketSetIt)->s < fdmIt->first) {
-            // On the other hand, if the iterator into socketSet is lower than the one into FDM, then we can discard
-            // the one in socketSet, it's not part of our candidate set.
-            socketSetIt++;
+
+        Socket* socket = nullptr;
+        {
+            // Lock and dequeue.
+            unique_lock<mutex> lock(server._networkMutex);
+
+            if (server._networkThreadQueue.size()) {
+                // If there's something in the queue, pull it off the queue.
+                auto first = server._networkThreadQueue.begin();
+                socket = *first;
+                server._networkThreadQueue.erase(first);
+            } else if (server._networkThreadShouldExit.load()) {
+                // If there's nothing in the queue, and we're supposed to exit, do that.
+                return;
+            } else {
+                // Otherwise there's nothing in the queue and we're not supposed to exit, so wait for work.
+                // We'll get interrupted if either work becomes available or we're supposed to exit, so just start at
+                // the top of the loop if that happens.
+                auto beforeWait = chrono::steady_clock::now();
+                server._networkCV.wait(lock);
+                auto afterWait = chrono::steady_clock::now();
+                waitTime += (afterWait - beforeWait);
+                continue;
+            }
         }
-    }
 
-    SINFO("Poll returned " << fdm.size() << " sockets to inspect and we accepted " << newSocketCount
-          << " new sockets. Total to inspect: " << sockets.size() << " of " << socketSet.size() << ".");
+        // The only way we can fall through the locked block above and get to here is by dequeuing something to work on, so
+        // we start on that here.
+        SINFO("Got work for socket " << socket);
```

Comment:
isn't this going to just print a memory address? seems like we could either make this dbug or change it to `socket->addr`, or `socket->id` which we log other places

Reply:
Yes. But this was here for testing and I should just remove it.